### PR TITLE
implement multi-MPPT inverter model

### LINKED ISSUE #457: Inverter with multiple MPPTs
It's been a couple years since I have used pvlib (and the progress is amazing) so forgive me if this has already been covered.  Is there an implementation of inverters with multiple MPPTs?

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes #457
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

An approach that leverages the existing function `pvlib.inverter.sandia` and the model parameters available in the SAM file. Conceptually, this is the same approach as in Pvsyst, although it appears that Pvsyst offers more options for adjusting the inverter's behavior (and Pvsyst uses an internally-developed inverter output model, which isn't fully documented in the public files).

### CODE REVIEW

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:185]: We should be careful about zip behavior and mismatched length. I suspect we either need to handle it with `itertools.zip_longest` or disallow it.

File: pvlib/tests/test_inverter.py
Comment (wholmgren) [inline pvlib/tests/test_inverter.py:100]: The implementation `test_sandia_multi` appears to allow for a variety of type inputs, so I think it's worth adding regression tests for float and array input. I can imagine several other implementations that would not be so accommodating.

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:183]: Any disadvantage to `power_dc = sum(p_dc)`?

```python
In [4]: import numpy as np

In [5]: a = np.array([1, 1])

In [6]: b = np.array([2, 2])

In [7]: sum((a, b))
Out[7]: array([3, 3])

In [8]: import pandas as pd

In [9]: c = pd.Series([1, 1])

In [10]: d = pd.Series([2, 2])

In [11]: sum((c, d))
Out[11]:
0    3
1    3
dtype: int64
```

File: pvlib/inverter.py
Comment (cwhanse) [reply pvlib/inverter.py:185]: v_dc and p_dc need to be paired, one entry per array. Entry could be float, Series or numeric.

If v_dc and p_dc are different length then I would raise an Error. OK?

Comment (wholmgren) [reply pvlib/inverter.py:185]: That works. I believe ValueError is most appropriate.

**Comment by mikofski** [comment]
I found these pages in the pvsyst help docs interesting:
* [Power sharing](https://www.pvsyst.com/help/powersharing.htm(
* Grid Inverters: main parameters] (https://www.pvsyst.com/help/inverters_general.htm)

>**Be carefu**l: as in the PVsyst inverter database, there is no information about the maximum power (or current) for each individual MPPT input, it is your responsibility to check in the data sheets whether the allocated power is compatible with the manufacturer's specifications.  PVsyst puts a  general limit to the sharing ratio, which is modifiable in the Hidden parameters.

Has anyone checked the _hidden parameters_?

**Comment by cwhanse** [comment]
> Has anyone checked the _hidden parameters_?

I don't have a Pvsyst license, I'd be interested to know what parameters are hidden that relate to multi-MPPT inverters.

I've read those pages and also exchanged email with Bruno at Pvsyst. I don't feel confident that I accurately understand how inverter efficiency is modeled in Pvsyst, so am not proposing a `inverter.pvsyst` function. 

Inverter datasheets typically list limits on DC current per MPP input, but these limits appear be related to safety (e.g. fuse ratings) rather than limits on DC-AC conversion.

For this model, I had proposed a parameter 'dc_limit' which would act as a cap on DC power at each MPP input. But after some discussions, I've removed that parameter for two, related reasons:
- there's no reliable source for its value (the safety limits in datasheets are not what is needed here)
- it is not clear that inverters actually impose limits per MPP input.

**Comment by cwhanse** [comment]
Bumping this PR as I think it's ready for review, and is needed to move #1076 forward.

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:152]: what's the use case for allowing numeric? Can we always require tuple/list of numeric?

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:182]: this will fail if a user provides a list instead of a tuple. I think most python users expect functions to treat lists and tuples as interchangeable. I'd argue a 1-d array should be acceptable in most cases, too.

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:195]: I don't understand why we'd include this line in the `sandia_multi` function. Isn't it duplicating functionality in `sandia`? Maybe I don't understand the use cases well enough.

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:48]: I was going to say the `np.where` in the original function does all of this in one line, but the conditional in the array case is different from the conditional in the float case, so I'm just confused.

File: pvlib/inverter.py
Comment (cwhanse) [reply pvlib/inverter.py:152]: to have this function behave appropriately when only one MPPT input is being used. In that case I'm not opposed to requiring the input to be `list or tuple of numeric` of length 1.

File: pvlib/inverter.py
Comment (cwhanse) [reply pvlib/inverter.py:195]: I was thinking that `inverter.sandia_multi(600., 1500., *params)` should work, but can't really say that's a required use.

Eventually I'd see `inverter.sandia` and `inverter.sandia_multi` merging, once the latter is better documented, with validation, and has broader acceptance.

File: pvlib/tests/test_inverter.py
Comment (cwhanse) [inline pvlib/tests/test_inverter.py:123]: @wholmgren do you think the use case in this test should be retained?

File: pvlib/inverter.py
Comment (cwhanse) [reply pvlib/inverter.py:48]: Line 47 is a typo, should be `p_dc < Pso:`.  The `try/except` is an attempt to handle the case when `power_ac` is a float. `np.where` converts `Series` to `ndarray`, and I was hoping to avoid another use of `isinstance` to preserve input type. I'm OK using the technique in `inverter.sandia` if you prefer that.

File: pvlib/tests/test_inverter.py
Comment (wholmgren) [reply pvlib/tests/test_inverter.py:123]: No

File: pvlib/tests/test_inverter.py
Comment (wholmgren) [inline pvlib/tests/test_inverter.py:119]: I'd also fail this unless changed to `...(vdcs,), (pdcs,)...`

File: pvlib/tests/test_inverter.py
Comment (cwhanse) [reply pvlib/tests/test_inverter.py:123]: That will simplify things.

File: pvlib/inverter.py
Comment (wholmgren) [reply pvlib/inverter.py:48]: > np.where converts Series to ndarray, and I was hoping to avoid another use of isinstance to preserve input type.

Ah, I forgot about that. Perhaps your method is preferable because it works with any array-like input (e.g. dask). I suggest assigning the condition and the limit value to variables before the try/except block to ensure there's no inconsistency.

**Comment by wholmgren** [comment]
@cwhanse you can decorate that failing psm test function with `@pytest.mark.xfail(strict=True, reason='github 1091')`

File: pvlib/inverter.py
Comment (wfvining) [reply pvlib/inverter.py:152]: +1 to requiring an Iterable of length 1

File: pvlib/inverter.py
Comment (cwhanse) [reply pvlib/inverter.py:182]: The problem with 1-d arrays is that this will run and produce garbage results:
```
inverter.sandia_multi(v_dc=np.array([500., 500.]), p_dc=np.array([1500., 1500.]), ...)
```
Because `p_dc` is iterable
```
power_dc = sum(p_dc)
```
sums over the array and produces a float instead of an array.

This doesn't work either since numpy doesn't create 1d arrays containing a single 1d array of any length:
```
inverter.sandia_multi(v_dc=np.array(np.array([500., 500.])), p_dc=np.array(np.array([1500., 1500.])), ...)
```
I'd just not list `array` as an accepted type for these inputs.

File: pvlib/inverter.py
Comment (cwhanse) [reply pvlib/inverter.py]: Do we want to test for and enforce `tuple` or `list` type? Duck-typing risks returning garbage results if the input type is array.

File: pvlib/inverter.py
Comment (wholmgren) [reply pvlib/inverter.py]: I'd hope it works with a 2D array

```
>>> np.array((np.array([500., 500.]), ))
array([[500., 500.]])
```

File: pvlib/inverter.py
Comment (wholmgren) [reply pvlib/inverter.py]: I'd prefer a garbage in, garbage out approach in this situation because it feels like we're still figuring out this function.

File: pvlib/inverter.py
Comment (wholmgren) [inline pvlib/inverter.py:183]: change to `!=` and put the `raise` in the `if` block, then move the rest of the code out of the `if/else` structure. no need for an `else`.

**Review by wholmgren — APPROVED** [review]

**Comment by cwhanse** [comment]
@pvlib/pvlib-maintainer will merge on 12/2 unless there are additional reviews

**Review by kanderso-nrel — APPROVED** [review]
LGTM, minor comments below

**Comment by cwhanse** [comment]
@pvlib/pvlib-maintainer Will merge today if no further reviews

**Comment by wholmgren** [comment]
let's move on with it - thanks @cwhanse!

**Comment by jhfatehi** [comment]
Sorry I am a bit late to the game on this.  My understanding of the multi-MPPT modeling method currently being worked on is that all arrays are operating at the same DC voltage.  The focus is address clipping per MPPT.  When I first brought up this question I was modeling a residential system with 2 arrays on different planes and of different sizes.  My solution at the time was to average the DC voltage between the 2 arrays and feed that into the inverter efficiency loss function.  Given the huge variations in insolation because of shading on a typical residential system, any accuracy loss because of the inverter efficiency was not a concern to me at the time.  No suggestion here.  I just wanted to circle back with the original use case.

**Pull request merged**
